Fix/wp get global styles for custom props returns internal variable

samnajian opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This PR fixes https://github.com/WordPress/gutenberg/issues/49693 so that `wp_get_global_styles` for the following in theme.json 
```
"core/post-terms": {
    "typography": { "fontSize": "var(--wp--preset--font-size--small)" },
    "color":{ "background": "var:preset|color|secondary" }
}
```
returns:

```
(
    [typography] => Array( [fontSize] => var(--wp--preset--font-size--small) )
    [color] => Array( [background] => var(--wp--preset--color--secondary) )
)
```

## Why?
https://github.com/WordPress/gutenberg/issues/49693
The internal syntax shouldn't leak out.

## How?
This PR extract the already existing logic that converts `var:preset|color|secondary` to `var(--wp--preset--font-size--small)` to a separate method, then uses the same method to sanitize the output of `wp_get_global_styles` to only include custom CSS variables and not internal variable syntax.

## Testing Instructions
Please refer to the issues for testing instructions, but please note the following:

1) Make sure the Gutenberg plugin is installed and active
2) In body of `wp_get_global_styles` exchange `WP_Theme_JSON_Resolver`  with `WP_Theme_JSON_Resolver_Gutenberg`
3) Confirm the output of `wp_get_global_styles` doesn't include internal variable syntax.

samnajian committed:
Extract logic that converts internal props to css variables

The logic that converts internal css variables to custom css
variables is extracted to a function

samnajian committed:
Convert the internal css variables to css custom variables

samnajian committed:
Add test to assert internal css syntax is converted to CSS variable

oandregal commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -2071,4 +2071,27 @@ public function data_process_blocks_custom_css() {
+				),
+			)
+		);
+		$styles = $result->get_data()['styles']['blocks']['core/post-terms'];
```
We can use `get_raw_data` which simply returns the internal theme.json structure and nothing else while `get_data` does a lot of other things:

```suggestion
		$styles = $result->get_raw_data()['styles']['blocks']['core/post-terms'];
```

oandregal commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -795,6 +795,7 @@ protected static function sanitize( $input, $valid_block_names, $valid_element_n
 			if ( empty( $result ) ) {
 				unset( $output[ $subtree ] );
 			} else {
+				$result             = static::sanitize_variables( $result, $schema[ $subtree ] );
```
We could get rid of the intermediate `$result` and do:

```php
$output[ $subtree ] = static::sanitize_variables( $result, $schema[ $subtree ] );
```

oandregal commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -1989,21 +1990,7 @@ protected static function get_property_value( $styles, $path, $theme_json = null
-		}
-
-		return $value;
+		return self::convert_custom_properties( $value );
```
If the constructor will take care of transforming data in the `var:preset|color|secondary` format to data in the `var(--wp--preset--color--secondary)` format, does any other code need to deal with that as well? Can we do instead?

```suggestion
		return $value;
```

samnajian replied:
Well this is a logic I just extracted to a separate method, and `get_property_value` is used in `compute_style_properties`, `get_layout_styles` and `get_root_layout_rules`, so I'd rather no change the output of `get_property_value` in this PR, apart from that since the conversion is added to `WP_Theme_JSON_Gutenberg::sanitize` I'm expecting the conversion to happen whenever the values are sanitized.

oandregal replied:
These are the "entry points" to this class:

- `__construct`: uses the `sanitize` method, which calls `sanitize_variables`.
- `remove_insecure_properties`: uses the `sanitize` method, which calls `sanitize_variables`.
- `get_from_editor_settings`: utility function, doesn't do anything with the internal `theme.json` dataset.
- `get_element_class_name`: utility function, doesn't do anything with the internal `theme.json` dataset.
- `scope_selector`: utility function, doesn't do anything with the internal `theme.json` dataset.

Any other method has the constructor or remove_insecure_properties as an ancestor. Given both sanitize the variables, I don't see how any other method can receive the internal format. Did I miss something? :thinking:

samnajian replied:
ah, I got your idea incorrectly the first time.

samnajian replied:
Just removed the [redundant usage](https://github.com/WordPress/gutenberg/pull/50366/commits/df276293029f46ee272e726181097a6a70a8f5bc) and everything seems to be fine, your analogy is correct since every method is using `$this->theme_json` which is already sanitized in the constructor.

oandregal commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -2071,4 +2071,27 @@ public function data_process_blocks_custom_css() {
+		$result = new WP_Theme_JSON_Gutenberg(
+			array(
+				'version' => WP_Theme_JSON_Gutenberg::LATEST_SCHEMA,
+				'styles'  => array(
```
In addition to block-level data, can we add test data for top-level, elements, and variations to cover the full space and recursion?

- top-level: `styles.color.background`
- block-level: `styles.blocks.core/post-terms.color.background`
- elements top-level: `styles.link.color.background`
- elements block-level: `styles.blocks.core/navigation.elements.link.color.background`
- style variations: `styles.blocks.core/quote.variations.plain.color.background`

samnajian replied:
Addressed in https://github.com/WordPress/gutenberg/pull/50366/commits/e8b1197ee3ca7ab8e51d187465aa7db8d90166df and https://github.com/WordPress/gutenberg/pull/50366/commits/adc1de945dae0a4ecfecca4dac9031fed76e7ac0

oandregal commented:
This is a great start, thanks!

> In body of wp_get_global_styles exchange WP_Theme_JSON_Resolver with WP_Theme_JSON_Resolver_Gutenberg

The only thing left would be to copy back from [core](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/global-styles-and-settings.php) the `wp_get_global_styles` function and declare it as `gutenberg_get_global_styles` function in Gutenberg. This is a bit cumbersome, but it's the method we have now to keep core<=>gutenberg in sync.

samnajian committed:
Define gutenberg_get_global_styles and update usages

As of the changes to WP_Theme_JSON_Resolver_Gutenberg,
this function had to get defined to keep in sync with core

samnajian committed:
Make sure variation styles are includes when sanitizing

oandregal commented on lib/compat/wordpress-6.3/get-global-styles-and-settings.php:
```diff
@@ -124,3 +124,35 @@ function wp_get_block_css_selector( $block_type, $target = 'root', $fallback = f
+/**
+ * Gets the styles resulting of merging core, theme, and user data.
+ *
+ * @since 6.3.0
```
This function is already in core:

```suggestion
 * @since 5.9.0
```

oandregal replied:
We can add a new line saying:

```php
@since 6.3.0 the internal link format "var:preset|color|secondary" is resolved to "var(--wp--preset--font-size--small)" so consumers don't have to.
```

samnajian committed:
Add correct since annotation to gutenberg_get_global_styles

Co-authored-by: André <583546+oandregal@users.noreply.github.com>

samnajian committed:
Add descriptive since annotation to gutenberg_get_global_styles

oandregal commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -3578,4 +3564,78 @@ protected function get_feature_declarations_for_node( $metadata, &$node ) {
+	 * @param string $value The variable such as var:preset|color|vivid-green-cyan to convert.
+	 * @return string The converted variable.
+	 */
+	protected static function convert_custom_properties( $value ) {
```
```suggestion
	private static function convert_custom_properties( $value ) {
```

oandregal commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -3578,4 +3564,78 @@ protected function get_feature_declarations_for_node( $metadata, &$node ) {
+	 * @param array $schema Schema to adhere to.
+	 * @return array The modified $tree.
+	 */
+	protected static function sanitize_variables( $tree, $schema ) {
```
```suggestion
	private static function sanitize_variables( $tree, $schema ) {
```

samnajian committed:
Remove usage of convert_custom_properties from get_property_value

Since convert_custom_properties is applied in constructor
the other methods don't need to apply it again

oandregal commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -3578,4 +3564,78 @@ protected function get_feature_declarations_for_node( $metadata, &$node ) {
+	 * @return array The modified $tree.
+	 */
+	private static function sanitize_variables( $tree, $schema ) {
+		$tree   = array_intersect_key( $tree, $schema );
```
:thinking: what is this for? Doesn't the tree already adhere to the schema (due to `remove_keys_not_in_schema` call)?

oandregal commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -3578,4 +3564,78 @@ protected function get_feature_declarations_for_node( $metadata, &$node ) {
+				continue;
+			}
+			$values = $tree[ $key ];
+			if ( is_array( $values ) ) {
```
What would you think of writing this as:

```php
if ( is_array( $values ) ) {
    // ...
} else if ( ! is_string( $values ) || 0 !== strpos( $values, $prefix ) ) {
    continue;
} else {
    $values = self::convert_custom_properties( $values );
}

oandregal replied:
Actually, the `sanitize_variables` function could be simplified as:

```php
	protected static function sanitize_variables( $tree, $schema ) {
		$prefix = 'var:';

		foreach ( $schema as $key => $data ) {
			if ( ! isset( $tree[ $key ] ) ) {
				continue;
			}
			$values = $tree[ $key ];
			if ( is_array( $values ) ) {
				$values = self::sanitize_variables( $tree[ $key ], $schema[ $key ] );
			} else if ( ! is_string( $values ) || 0 !== strpos( $values, $prefix ) ) {
				continue;
			} else {
				$values = self::convert_custom_properties( $values );
			}

			$tree[ $key ] = $values;
		}

		return $tree;
	}
```

samnajian replied:
Thanks for the suggestion, it's applied, but in `$values = self::sanitize_variables( $tree[ $key ], $schema[ $key ] );`  won't work as parts of the data may still need the whole `$schema` and not only ` $schema[ $key ] `

oandregal commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -3578,4 +3564,78 @@ protected function get_feature_declarations_for_node( $metadata, &$node ) {
+						$values[ $name ] = array_merge( $value, self::sanitize_variables( $value, $schema ) );
+						continue;
+					}
+					if ( ! is_string( $value ) || 0 !== strpos( $value, $prefix ) ) {
```
Sharing what I learn. Looking at `0 !== strpos( $value, $prefix )` I wondered whether the `var:` prefix could be found in any position other than the first characters. The [old code](https://github.com/WordPress/gutenberg/pull/50366/files#diff-b03597cc3da199e5aa5a94950e8241135904853e7c3b82d758e42744878afae7L1997) used `if ( 0 === strncmp( $value, $prefix, $prefix_len ) )` so it only supported `var:` in the first characters. Using `0 !== strpos` here should be fine then.

oandregal commented:
I've left a few suggestions for improving `sanitize_variables` implementation. Other than that, this is ready. Thanks for your work here :clap:

samnajian committed:
Apply improvements suggested in PR review

samnajian committed:
Simplify sanitize_variables and update usage

Since the data is already checked against the schema
sanitize_variables doesn't need to know about the schema
anymore

samnajian committed:
Rename sanitize_variables to resolve_custom_css_format

oandregal reviewed (approved):
This is great work, thanks for your contributions, Sam!

hellofromtonya commented on lib/compat/wordpress-6.3/get-global-styles-and-settings.php:
```diff
@@ -124,3 +124,37 @@ function wp_get_block_css_selector( $block_type, $target = 'root', $fallback = f
+ * Gets the styles resulting of merging core, theme, and user data.
+ *
+ * @since 5.9.0
+ * @since 6.3.0 the internal link format "var:preset|color|secondary" is resolved
```
Once introduced in WP Core, this modifies the value format for the returned styles from `wp_get_global_styles()`. As such, it is a backwards-compatibility (BC) break for extenders expecting and using this structure.

The result could mean broken styling experiences for users.

Is this a truly a bugfix, meaning `wp_get_global_styles()` should always have returned the style value structure? Or is this a new improvement to avoid leaking internal structures?

cc @oandregal

oandregal replied:
This is why I think it's a fine change.

Let say we had the following input:

```php
"core/post-terms": {
    "typography": { "fontSize": "var(--wp--preset--font-size--small)" },
    "color":{ "background": "var:preset|color|secondary" }
}
```

**Before** this PR, the output of `wp_get_global_styles` would be:

```php
(
    [typography] => Array( [fontSize] => var(--wp--preset--font-size--small) )
    [color] => Array( [background] => var:preset|color|secondary )
)
```

**After** this PR, the output of `wp_get_global_styles` would be:

```php
(
    [typography] => Array( [fontSize] => var(--wp--preset--font-size--small) )
    [color] => Array( [background] => var(--wp--preset--color--secondary) )
)
```

Before this PR, the consumers had to deal with **two formats**: `var(--wp--preset--color--secondary)` and `var:preset|color|secondary`, as a possible representation of the same value.

After this PR, they only have to deal with **one format** that's been always present: `var(--wp--preset--color--secondary)`. I don't remember when the `var:preset|color|secondary` was introduced for people to use in `theme.json` but it was later and added as a convenience. It should have never leaked to consumers, in my view. I did some quick search but didn't find when it was, I can do some archeology if this is important.

An important note is that the format doesn't carry any meaning about _where_ this data comes from. Any origin (default, block, theme, user) can use `var:preset|color|secondary`. Consumers cannot use that format as a hint of it being data from the user origin.

Does this help?

oandregal commented:
Follow-up to fix PHP docblock at https://github.com/WordPress/gutenberg/pull/50527

oandregal commented:
Backport at https://github.com/WordPress/wordpress-develop/pull/4556
